USWDS - Button: Add basic support for usa-icon in buttons

mejiaj opened the pull request:
# Summary

New features:
- **Improved icon alignment in buttons.** Buttons that include an inline `usa-icon` are now visually centered.
- **Improved control of inner content.** Buttons use flexbox for vertical alignment and column spacing. Added a new setting `$theme-button-icon-gap: 1 !default;` that takes units for inner content.
- **New Button controls in Storybook.** There are two new controls for internal testing: `add_icon` & `icon_name`.

    <details><summary><b>Control: <code>add_icon</code></b></summary>
    <p></p>
    
    A toggle for including an icon in stories. 
    <img width="807" alt="image" src="https://github.com/uswds/uswds/assets/3385219/d50c2e44-32c6-4a4b-b488-66d5e1472c15">

    
    </details> 
    
    <details><summary><b>Control: <code>icon_name</code></b>  (Depends on <code>add_icon</code>)</summary>
    <p></p>
    
   

    A dropdown of all USA icons with `add_circle_outline` as the default. Icon was mentioned in the original issue and was used as the basis for styling.
    
    <img width="807" alt="image" src="https://github.com/uswds/uswds/assets/3385219/f0d7b02e-9536-4621-a33e-8e3a71c3115e">
    
    </details> 


## Breaking change

This is not a breaking change.

## Related issue

Closes #5359.

<!--
Every pull request should resolve an open issue.
If no open issue exists, you can open one here:
https://github.com/uswds/uswds/issues/new/choose.
-->

## Related pull requests

- [Changelog entry](https://github.com/uswds/uswds-site/pull/2207)
- Related to #5657. Fixes alignment issue mentioned in PR.

## Preview links

- [Button with icon](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/jm-enhancement-button-icons/?path=/story/components-button--icon).
- [Button accent cool variant](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/jm-enhancement-button-icons/?path=/story/components-button--accent-cool). _Shows new controls available to button component._
- [Develop branch preview](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/develop/?path=/story/components-button--default)

## Problem statement

USA Icons in buttons weren't centered vertically.

**Before**
<img width="800" alt="image" src="https://github.com/uswds/uswds/assets/3385219/126bc517-8896-46eb-b784-52261b1bb758">

**After**

Added icon to both sides to illustrate spacing via `$theme-button-icon-gap`.

<img width="800" alt="image" src="https://github.com/uswds/uswds/assets/3385219/0e97648c-9073-4efa-be0e-5d8033f05b2f">


## Solution

Improved support for buttons with inline `usa-icons` using `inline-flex`.

### Why this approach was chosen

1. Improved control over inner spacing via the `column-gap` property.
2. Icons before/after will be properly spaced.
3. The move from `inline-block → inline-flex` doesn't seem like a big departure in display.

~Follows a similar pattern to what's been done in [breadcrumb:14](https://github.com/uswds/uswds/blob/7fa6bdae8629c0ee90a3f6405db089622fb4d0cd/packages/usa-breadcrumb/src/styles/_usa-breadcrumb.scss#L14) & [external link:32-40](https://github.com/uswds/uswds/blob/7fa6bdae8629c0ee90a3f6405db089622fb4d0cd/packages/uswds-core/src/styles/mixins/general/external-link.scss#L32-L40). A small enhancement seemed less heavy handed than other solution.~

### How I implemented

Switched from `inline-block → inline-flex`. Added properties for column gap and justify content (for centering text).

~Added vertical alignment and negative margin to visually center.~

### Alternatives
Using the `:has()` selector to use `inline-grid` ~or `inline-flex`~. These features provide more control, but have their own problems:  

1. `:has()` doesn't have crossbrowser support.
4. There are conflicts with postcsso and `:has()` pseudoselector. The plugin for postcss requires a JS polyfill to be included & has performance issues.

<!--
It can be helpful if we understand:
1. What the solution is,
6. Why this approach was chosen,
7. How you implemented the change, and
8. Possible limitations of this approach and alternate solution paths.
-->

## Major changes

| Variable | Default | Type | Description |
|:--------|--------|:--------|:--------|
| `$theme-button-icon-gap` | 1 | units | The inner flexbox column gap size. | 

## Testing and review

#### Icon alignment

1. Visit [button icon preview](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/jm-enhancement-button-icons/?path=/story/components-button--icon) link.
5. Confirm icon is visually centered.

#### Storybook Controls

1. Go to any button variant, like [Accent warm](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/jm-enhancement-button-icons/?path=/story/components-button--accent-warm).
2. Confirm new control `add_icon` is visible.
6. Toggle to `True`.
7. Confirm that new control `icon_name` shows as a dropdown for setting icon.
8. Confirm that all USA icons are available as options in dropdown.
9. Confirm that icon changes on selection.

### Testing checklist

- [x] Icons in buttons should be visually centered.
- [x] Button stories have a control to toggle icons.
    - [x] If `Include icon` is true, show a dropdown of icons.
    - [x] Icons should update on select.
- [x] There should be no `html:build` errors.

<!--
## Dependency updates

| Dependency name              | Previous version | New version |
| ---------------------------- | :--------------: | :---------: |
| [Updated dependency example] |     [1.0.0]      |   [1.0.1]   |
| [New dependency example]     |        --        |   [3.0.1]   |
| [Removed dependency example] |     [2.10.2]     |     --      |
-->
<!--
For PRs that include dependency updates, uncomment this section and
include a list of the changed dependencies and version numbers.
-->

<!--
Before opening this PR, make sure you’ve done whichever of these applies to you:
- [ ] Confirm that this code follows the [18F Front End Coding Style Guide](https://pages.18f.gov/frontend/) and [Accessibility Guide](https://pages.18f.gov/accessibility/checklist/).
- [ ] Run `git pull origin [base branch]` to pull in the most recent updates from your base and check for merge conflicts. (Often, the base branch is `develop`).
- [ ] Run `npm run prettier:sass` to format any Sass updates.
- [ ] Run `npm test` and confirm that all tests pass.
- [ ] Run your code through [HTML_CodeSniffer](http://squizlabs.github.io/HTML_CodeSniffer/) and make sure it’s error free.
-->

mahoneycm commented on packages/usa-button/src/usa-button.twig:
```diff
@@ -2,20 +2,80 @@
+  </button>
 {% else %}
-  <button type="{{ type }}" class="usa-button {{ modifier }}">{{ text }}</button>
+  <button type="{{ type }}" class="usa-button {{ modifier }}">
+    {{ text }}
+  </button>
```
I'd be interested in being able to see the icon even when `is_demo` is false.

![image](https://github.com/uswds/uswds/assets/25211650/a4d48126-6bab-4082-8d70-f8f9b420a828)


```suggestion
  <button type="{{ type }}" class="usa-button {{ modifier }}">
    {{ text }}
    {% if icon_included %}
    {{ icon }}
    {% endif %}
  </button>
```

mejiaj replied:
Good point. Created a macro for more consistent display in 87fc766. 


_We still need #5394 to clean this up further._

mahoneycm commented on packages/usa-button/src/usa-button.twig:
```diff
@@ -2,20 +2,80 @@
-  <button type="{{ type }}" class="usa-button {{ modifier }} usa-button--unstyled">Unstyled button</button>
+  <button type="{{ type }}" class="usa-button {{ modifier }}">
+    Default
+    {% if icon_included %}
+    {{ icon }}
+    {% endif %}
```
Should we indent these for improved legibility? 
```suggestion
    {% if icon_included %}
       {{ icon }}
    {% endif %}
```

>**Warning**
> Github suggestion won't let me hit the tab button so this is just a few spaces. Don't commit this example.

mejiaj replied:
## Indents

Resolving for now since the indent note doesn't apply anymore - Reworked in 87fc766.

## Spacing

@amyleadem this might be okay for now. Open to discussing further in a separate enhancement. We'd have to solution how to target styles when icon is placed before/after text.

mahoneycm reviewed (changes requested):
Diggin' this one! Excited for the potential addition to the `copy-code` button!

I personally would like to still be able to view icons when `is_demo` is false. Let me know if you disagree!

## Testing checklist

- [x] Icons in buttons should be visually centered.
- [x] Button stories have a control to toggle icons.
    - [x] If `Include icon` is true, show a dropdown of icons.
    - [x] Icons should update on select.
- [x] There should be no `html:build` errors.

amyleadem reviewed (changes requested):
@mejiaj I really like how you set up the Storybook controls here. The conditional control is great and importing the icon data was a really nice touch. 

I added a couple of notes in the comments. Mostly, I am curious if it is possible to use flexbox for button alignment. I was looking at https://github.com/uswds/uswds/pull/5324 about text wrapping in button groups at the same time and it made me curious if there was a solution that would work for both issues. I added more detail in the comments. 

There is a chance I am missing something, so I'm curious to hear what you think. Please let me know if you have any questions.

mejiaj commented on packages/usa-button/src/usa-button.stories.js:
```diff
@@ -31,6 +36,20 @@ export default {
       control: { type: "radio" },
     },
+    icon_included: {
+      name: "Icon included",
```
Good idea. Updated in aacb83b.

mejiaj commented on packages/usa-button/src/usa-button.stories.js:
```diff
@@ -51,6 +70,13 @@ Base.args = BaseContent;
+Icon.args = {
+  ...DefaultContent,
+  icon_included: true,
+  icon_name: "add_circle_outline",
```
Explicitly setting it will preselect it in the control, which I think is a good feature. Otherwise it will **still** show the icon, but it won't be selected. Added a note in 3c3684e.

### With `icon_name`
<img width="500" alt="image" src="https://github.com/uswds/uswds/assets/3385219/c13b2813-392d-46cf-821d-fae4cc563284">


### Without `icon_name`
<img width="500" alt="image" src="https://github.com/uswds/uswds/assets/3385219/8aa3214c-1ae2-49a6-a862-cb9e574c173b">

mejiaj commented on packages/usa-button/src/styles/_usa-button.scss:
```diff
@@ -52,6 +52,13 @@ $button-stroke: inset 0 0 0 units($theme-button-stroke-width);
     @include button-disabled;
   }
 
+  // @TODO: Use `:has()` selector once browser support improves.
+  // https://developer.mozilla.org/en-US/docs/Web/CSS/:has#browser_compatibility
+  .usa-icon {
+    margin-top: -0.1em; // Similar pattern to _usa-breadcrumb.scss:14.
+    vertical-align: middle;
+  }
+
```
@amyleadem 

Re: display: `inline-flex`. Because this is such a widely used button ― I wanted to minimize any potentially breaking changes as much as possible. 

Since we already specify `inline-block`, I don't think the switch to `inline-flex` is as drastic as I imagined. Double-checking MDN [display - CSS: Cascading Style Sheets | MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/display#grouped_values) and there should be very little display difference at all.

This addition improves the experience, but might require a new theme setting for the column gap. I'll play with this.

mejiaj commented on packages/uswds-core/src/styles/mixins/general/button-unstyled.scss:
```diff
@@ -10,9 +10,10 @@
   box-shadow: none;
   font-weight: font-weight("normal");
+  justify-content: normal;
+  text-align: left;
```
Added text-align back in to fix a regression with Headers nav link (uses button & accordion) 

**Regression**
<img width="253" alt="image" src="https://github.com/uswds/uswds/assets/3385219/5218015f-4918-4988-8101-cd267925d646">

mahoneycm replied:
Just curious, should this be defined in Header instead since that's where it's relevant?

mejiaj replied:
@mahoneycm since this line was previously removed and added back in. Moving it to header might break other uses of the mixin.

mahoneycm replied:
Thanks for clarifying!

mahoneycm reviewed (changes requested):
Looks good to me! One quick question about the PR description:

 Is this note from your PR still correct? Seems like it was from your previous implementation
        
> ### Why this approach was chosen
> 
> Follows a similar pattern to what's been done in [breadcrumb:14](https://github.com/uswds/uswds/blob/7fa6bdae8629c0ee90a3f6405db089622fb4d0cd/packages/usa-breadcrumb/src/styles/_usa-breadcrumb.scss#L14) & [external link:32-40](https://github.com/uswds/uswds/blob/7fa6bdae8629c0ee90a3f6405db089622fb4d0cd/packages/uswds-core/src/styles/mixins/general/external-link.scss#L32-L40). A small enhancement seemed less heavy handed than other solution.
> 
> ### How I implemented
> 
> Added vertical alignment and negative margin to visually center.
> 

---

## Review checklist

- [x]  Icons in buttons should be visually centered.
- [x]  Button stories have a control to toggle icons.
    - [x]  If `Include icon` is true, show a dropdown of icons.
    - [x]  Icons should update on select.
- [x]  There should be no `html:build` errors.
- [x]  Code looks good
- [x]  No visual regressions where <button class="usa-button"></button>` is used
- [x]  Tested on site with Copy Code button and it worked like a charm 👍 
- [x]  Template w/ Macros still easily modifiable
- [ ]  PR description is up to date

mejiaj commented:
@mahoneycm thanks for the review. I've responded to questions & updated PR description ― good catch!

amyleadem reviewed (changes requested):
This is looking good! I had just a few notes:

- I like the new setting but was initially confused about its purpose based solely on the name. Wondering if it might be clearer to name`$theme-button-column-gap` something like `$theme-button-icon-gap`
- In the card component, I noticed that the text was aligning to the left in the button when the line broke (Details in comment below).
- Some PR description things:
    - In the PR description, the control is referenced as “Icon included” but in storybook it is called “Add icon”
    - I wonder if we should flag that this as “potentially breaking” breaking because it might adjust visual presentation if users have already added inline svgs to their buttons

Here are the things I tested in Chrome, Safari, and Firefox: 

- [x] Confirm that new storybook controls work as expected
    - [x] Setting “Add icon” to `true` adds an icon to the button and adds a dropdown control to select the icon
    - [x] Changing the icon in the “icon name” control changes the icon in the button
- [x] Confirm that icons are vertically centered:
    - [x] with a single line of text
    - [x] with multiple lines of text
    - [x] with large font sizes
    - [x] with small font sizes
    - [x] with large icon sizes
    - [x] with small icon sizes
- [x] Confirm that icons are aligned when added before the text
- [x] Confirm that updating the value in `$theme-button-column-gap` changes the space between buttons as expected
- [x] Confirm no visual regressions in button or button group components
- [ ] Confirm that button groups are aligned in height when only some buttons in the group have an icon
    - Adding an icon to one button in a group makes the button a different height. I think this is ok but it would be nice to do a fast follow-up to fix this.    
        ![image](https://github.com/uswds/uswds/assets/93996430/470e4f54-0d92-4e9a-a4f8-66d5562251ce)

mahoneycm commented:
I like @amyleadem's idea of renaming the setting from `$theme-button-column-gap` to something like `$theme-button-icon-gap`. It's a bit more descriptive which I appreciate!

mejiaj commented:
@amyleadem @mahoneycm appreciate the suggestions. I've made the following changes:

- [x] Renamed icon setting to `$theme-button-column-gap → $theme-button-icon-gap`.
- [x] Update PR description.
- [x] Added `text-align: center` back to button.
- [x] Internal: Renamed storybook control `icon_included → add_icon`.

amyleadem reviewed (approved):
Looks good to me! 

All updates look good. I also was able to successfully test both the visuals and settings configuration for `$theme-button-icon-gap` in uswds-sandbox.


```html
<button class="usa-button" type="button">
    Default
    <svg class="usa-icon" aria-hidden="true" focusable="false" role="img">
        <use xlink:href="/assets/img/sprite.svg#add"></use>
      </svg>
</button>
```
`$theme-button-icon-gap: 1`

![image](https://github.com/uswds/uswds/assets/93996430/f1d47b49-33e7-41d9-b26f-faee1f55ea78)

`$theme-button-icon-gap: 3`

![image](https://github.com/uswds/uswds/assets/93996430/c395707d-fdf7-4f11-9410-52668d6082dc)

amyleadem commented:
@mejiaj One more question - should we consider this a potentially breaking change since users might have already added custom/conflicting styles to add icons? Probably worth warning them to confirm that their buttons with icons display as expected.

mahoneycm reviewed (approved):
## Review checklist

- [x]  Icons in buttons should be visually centered.
- [x]  Button stories have a control to toggle icons.
    - [x]  If `Include icon` is true, show a dropdown of icons.
    - [x]  Icons should update on select.
- [x] `$theme-button-icon-gap` works as expected. 
- [x]  There should be no `html:build` errors.
- [x]  Code looks good
- [x]  No visual regressions where `<button class="usa-button"></button>` is used
- [x]  Tested on site with Copy Code button and it worked like a charm 👍 
- [x]  Template w/ Macros still easily modifiable
- [x]  PR description is up to date

mejiaj commented:
> @mejiaj One more question - should we consider this a potentially breaking change since users might have already added custom/conflicting styles to add icons? Probably worth warning them to confirm that their buttons with icons display as expected.

Undecided on this. I don't feel like the impact is that big, but open to suggestions. 

It would be worth flagging that it's now supported so users _can_ remove their custom styles. But labelling it `Potentially breaking` seems heavy.

mahoneycm commented on packages/usa-button/src/styles/_usa-button.scss:
```diff
@@ -11,14 +11,16 @@ $button-stroke: inset 0 0 0 units($theme-button-stroke-width);
+  justify-content: center;
   margin-right: units(1);
   padding: units(1.5) units(2.5);
-  text-align: center;
```
Just noting this appears to be resolved 👍

thisisdano reviewed (approved):
Good update
